evaluation.md

```diff
@@ -0,0 +1,413 @@
+    .map(test_recommenders)\
+    .map(lambda x: (x, 1))\
+    .reduceByKey(add)\
+    .collect()
+```
+
+    CPU times: user 9.76 s, sys: 632 ms, total: 10.4 s
+    Wall time: 25min 35s
+
+
+
+```python
+results = defaultdict(int, results)
+```
+
+
+```python
+num_clients = sum(results.values())
+```
+
+## Computing individual counts
+
+
+```python
+individual_counts = []
+
+for i in range(len(recommenders)):
+    count = 0
+    
+    for key, key_count in results.items():
+        if key[i]:
+            count += key_count
+            
+    individual_counts.append(count)
+```
+
+## Displaying the results
+
+
+```python
+from pandas import DataFrame
+```
+
+
+```python
+def format_int(num):
+    return "{:,}".format(num)
+```
+
+
+```python
+def format_frequency(frequency):
+    return "%.3f%%" % (frequency  * 100)
+```
+
+
+```python
+def count_summary(counts):
+    absolute = map(format_int, counts)
+    relative = [format_frequency(count / float(num_clients)) for count in counts]
+    
+    return zip(absolute, relative)
+```
+
+This is a bit hacky. Sorting a data frame by formatted counts does not work; so we have to add the unformatted ones, sort the data frame, and then remove that column again.
+
+
+```python
+def sorted_dataframe(df, order, key="unformatted_counts"):
+    df[key] = order
+    return df.sort_values(by=key, ascending=False).drop(key, axis=1)
+```
+
+### Individual counts
+
+
+```python
+df = DataFrame(index=recommenders.keys(),
+          columns=["Absolute", "Relative"],
+          data=count_summary(individual_counts)
+)
+
+sorted_dataframe(df, individual_counts)
+```
+
+
+
+
+<div>
+<table border="1" class="dataframe">
+  <thead>
+    <tr style="text-align: right;">
+      <th></th>
+      <th>Absolute</th>
+      <th>Relative</th>
+    </tr>
+  </thead>
+  <tbody>
+    <tr>
+      <th>locale</th>
+      <td>5,433,935</td>
+      <td>89.227%</td>
+    </tr>
+    <tr>
+      <th>collaborative</th>
+      <td>5,202,700</td>
+      <td>85.430%</td>
+    </tr>
+    <tr>
+      <th>similarity</th>
+      <td>1,676,004</td>
+      <td>27.520%</td>
+    </tr>
+    <tr>
+      <th>legacy</th>
+      <td>93,968</td>
+      <td>1.543%</td>
+    </tr>
+  </tbody>
+</table>
+</div>
+
+
+
+$\implies$ The locale and collaborative recommenders are able to generate recommendations most of the time. The legacy recommender can only make recommendations very seldomly as not many users seem to have (legacy) addons installed.
+
+### Combined counts
+
+It's interesting to see how well the individual recommenders complement each other. In the following, we count how often different combinations of the recommenders can make recommendations.
+
+
+```python
+counts = count_summary(results.values())
+data = [elems + counts for elems, counts in zip(results.keys(), counts)]
+```
+
+
+```python
+df = DataFrame(columns=recommenders.keys() + ["Absolute", "Relative"],
+          data=data
+)
+
+sorted_dataframe(df, results.values())
```

Comment:
Please separate the rows where **only** one recommender is capable of making a recommendation. These results would be a lot easier to interpret is the tables were grouped by the number of `True` values in the row.